src/client/datascience/context/telemetry.ts

```diff
@@ -0,0 +1,304 @@
+                language = PYTHON_LANGUAGE;
+                break;
+            default:
+                break;
+        }
+        currentData.kernelLanguage = getTelemetrySafeLanguage(language);
+        // Keep track of the kernel that was last selected.
+        context.previouslySelectedKernelConnectionId = getKernelConnectionId(kernelConnection);
+
+        const interpreter = kernelConnection.interpreter;
+        if (interpreter) {
+            currentData.isUsingActiveInterpreter = WorkspaceInterpreterTracker.isActiveWorkspaceInterpreter(
+                resource,
+                interpreter
+            );
+            currentData.pythonEnvironmentType = interpreter.envType;
+            currentData.pythonEnvironmentPath = hashjs.sha256().update(interpreter.path).digest('hex');
+            if (interpreter.version) {
+                const { major, minor, patch } = interpreter.version;
+                currentData.pythonEnvironmentVersion = `${major}.${minor}.${patch}`;
+            } else {
+                currentData.pythonEnvironmentVersion = undefined;
+            }
+        }
+
+        currentData.kernelConnectionType = currentData.kernelConnectionType || kernelConnection?.kind;
+    } else {
+        context.previouslySelectedKernelConnectionId = '';
+    }
+    trackedInfo.set(key, [currentData, context]);
+}
+export function deleteTrackedInformation(resource: Uri) {
+    trackedInfo.delete(getUriKey(resource));
+}
+
+function getUriKey(uri: Uri) {
+    return currentOSType ? uri.fsPath.toLowerCase() : uri.fsPath;
+}
+
+function getContextualPropsForTelemetry(resource: Resource): ResourceSpecificTelemetryProperties | undefined {
+    if (!resource) {
+        return;
+    }
+    const data = trackedInfo.get(getUriKey(resource));
+    const resourceType = getResourceType(resource);
+    if (!data && resourceType) {
+        return {
+            resourceType
+        };
+    }
+    return data ? data[0] : undefined;
+}
+/**
+ * Some information such as interrupt counters & restart counters need to be reset
+ * after we have successfully interrupted or restarted a kernel.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function resetData(resource: Resource, eventName: string, properties: any) {
+    // Once we have successfully interrupted, clear the interrupt counter.
+    if (eventName === Telemetry.NotebookInterrupt) {
+        let kv: Pick<IEventNamePropertyMapping, Telemetry.NotebookInterrupt>;
+        const data: typeof kv[Telemetry.NotebookInterrupt] = properties;
+        // Check result to determine if success.
+        if ('result' in data && data.result === InterruptResult.Success) {
+            clearInterruptCounter(resource);
+        }
+    }
+    // Once we have successfully restarted, clear the interrupt counter.
+    if (eventName === Telemetry.NotebookRestart) {
+        let kv: Pick<IEventNamePropertyMapping, Telemetry.NotebookInterrupt>;
```

Comment:
This all looks incorrect in this block. Looks like it should be using restart methods and events? Copy paste error?